File: app/src/main/java/com/afollestad/photoaffix/dialogs/ImageSpacingDialog.java
package com.afollestad.photoaffix.dialogs;

import android.app.Activity;
import android.app.Dialog;
import android.app.DialogFragment;
import android.graphics.Color;
import android.os.Bundle;
import android.support.v4.content.ContextCompat;
import android.view.View;
import android.widget.SeekBar;
import android.widget.TextView;

import com.afollestad.materialdialogs.MaterialDialog;
import com.afollestad.materialdialogs.internal.MDTintHelper;
import com.afollestad.photoaffix.R;
import com.afollestad.photoaffix.ui.MainActivity;
import com.afollestad.photoaffix.utils.Prefs;
import com.afollestad.photoaffix.views.LineView;

import butterknife.BindView;
import butterknife.ButterKnife;
import butterknife.Unbinder;

/** @author Aidan Follestad (afollestad) */
public class ImageSpacingDialog extends DialogFragment {

  @BindView(R.id.spacingHorizontalSeek)
  SeekBar horizontalSeek;
  @BindView(R.id.spacingHorizontalLabel)
  TextView horizontalLabel;
  @BindView(R.id.spacingVerticalSeek)
  SeekBar verticalSeek;
  @BindView(R.id.spacingVerticalLabel)
  TextView verticalLabel;
  @BindView(R.id.verticalLine)
  LineView verticalLine;
  @BindView(R.id.horizontalLine)
  LineView horizontalLine;
  private SpacingCallback context;
  private Unbinder unbinder;

  public ImageSpacingDialog() {}

  @Override
  public Dialog onCreateDialog(Bundle savedInstanceState) {
    int fillColor = Prefs.bgFillColor(getActivity());
    if (fillColor == Color.TRANSPARENT)
      fillColor = ContextCompat.getColor(getActivity(), R.color.colorAccent);
    MaterialDialog dialog =
        new MaterialDialog.Builder(getActivity())
            .title(R.string.image_spacing)
            .customView(R.layout.dialog_imagespacing, true)
            .positiveText(android.R.string.ok)
            .negativeText(android.R.string.cancel)
            .positiveColor(fillColor)
            .negativeColor(fillColor)
            .widgetColor(fillColor)
            .onPositive((materialDialog, dialogAction) -> notifyActivity())
            .onNegative(
                (materialDialog, dialogAction) -> MainActivity.dismissDialog(materialDialog))
            .build();

    final View v = dialog.getCustomView();
    assert v != null;
    unbinder = ButterKnife.bind(this, v);

    SeekBar.OnSeekBarChangeListener seekListener =
        new SeekBar.OnSeekBarChangeListener() {
          @Override
          public void onProgressChanged(SeekBar seekBar, int progress, boolean fromUser) {
            if (seekBar.getId() == R.id.spacingHorizontalSeek) {
              horizontalLabel.setText(String.format("%d", progress));
              horizontalLine.setWidth(progress);
            } else {
              verticalLabel.setText(String.format("%d", progress));
              verticalLine.setWidth(progress);
            }
          }

          @Override
          public void onStartTrackingTouch(SeekBar seekBar) {}

          @Override
          public void onStopTrackingTouch(SeekBar seekBar) {}
        };
    horizontalSeek.setOnSeekBarChangeListener(seekListener);
    verticalSeek.setOnSeekBarChangeListener(seekListener);

    MDTintHelper.setTint(horizontalSeek, fillColor);
    MDTintHelper.setTint(verticalSeek, fillColor);
    horizontalLine.setColor(fillColor);
    verticalLine.setColor(fillColor);

    final int[] spacing = Prefs.imageSpacing(getActivity());
    horizontalSeek.setProgress(spacing[0]);
    verticalSeek.setProgress(spacing[1]);

    return dialog;
  }

  @Override
  public void onDestroyView() {
    super.onDestroyView();
    unbinder.unbind();
    unbinder = null;
  }

  private void notifyActivity() {
    if (context == null) return;
    context.onSpacingChanged(horizontalSeek.getProgress(), verticalSeek.getProgress());
  }

  @Override
  public void onAttach(Activity activity) {
    super.onAttach(activity);
    context = (SpacingCallback) activity;
  }

  public interface SpacingCallback {

    void onSpacingChanged(int horizontal, int vertical);
  }
}

File: app/src/main/java/com/afollestad/photoaffix/dialogs/ImageSpacingDialog.kt
/*
 * Licensed under Apache-2.0
 *
 * Designed and developed by Aidan Follestad (@afollestad)
 */
package com.afollestad.photoaffix.dialogs

import android.app.Dialog
import android.content.Context
import android.graphics.Color.TRANSPARENT
import android.os.Bundle
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat.getColor
import androidx.fragment.app.DialogFragment
import com.afollestad.materialdialogs.MaterialDialog
import com.afollestad.materialdialogs.customview.customView
import com.afollestad.materialdialogs.customview.getCustomView
import com.afollestad.photoaffix.App
import com.afollestad.photoaffix.R
import com.afollestad.photoaffix.di.BgFillColor
import com.afollestad.photoaffix.di.ImageSpacingHorizontal
import com.afollestad.photoaffix.di.ImageSpacingVertical
import com.afollestad.photoaffix.utils.onProgressChanged
import com.afollestad.rxkprefs.Pref
import kotlinx.android.synthetic.main.dialog_imagespacing.view.horizontalLine
import kotlinx.android.synthetic.main.dialog_imagespacing.view.spacingHorizontalLabel
import kotlinx.android.synthetic.main.dialog_imagespacing.view.spacingHorizontalSeek
import kotlinx.android.synthetic.main.dialog_imagespacing.view.spacingVerticalLabel
import kotlinx.android.synthetic.main.dialog_imagespacing.view.spacingVerticalSeek
import kotlinx.android.synthetic.main.dialog_imagespacing.view.verticalLine
import javax.inject.Inject

interface SpacingCallback {

  fun onSpacingChanged(
    horizontal: Int,
    vertical: Int
  )
}

/** @author Aidan Follestad (afollestad) */
class ImageSpacingDialog : DialogFragment() {

  companion object {
    private const val TAG = "[IMAGE_SIZING_DIALOG]"

    fun show(context: AppCompatActivity) {
      ImageSpacingDialog().show(context.supportFragmentManager, TAG)
    }
  }

  private var context: SpacingCallback? = null

  @Inject
  @field:BgFillColor
  lateinit var bgFillColorPref: Pref<Int>
  @Inject
  @field:ImageSpacingVertical
  lateinit var verticalSpacingPref: Pref<Int>
  @Inject
  @field:ImageSpacingHorizontal
  lateinit var horizontalSpacingPref: Pref<Int>

  override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
    val myActivity = activity
    requireNotNull(myActivity)

    (myActivity.applicationContext as App)
        .appComponent
        .inject(this)

    var fillColor = bgFillColorPref.get()
    if (fillColor == TRANSPARENT) {
      fillColor = getColor(myActivity, R.color.colorAccent)
    }

    val dialog = MaterialDialog(myActivity)
        .title(R.string.image_spacing)
        .customView(R.layout.dialog_imagespacing, scrollable = true)
        .positiveButton(android.R.string.ok) { notifyActivity() }
        .negativeButton(android.R.string.cancel) { dismiss() }

    val customView = dialog.getCustomView() ?: return dialog

    customView.spacingHorizontalSeek.max = 100
    customView.spacingHorizontalSeek.onProgressChanged {
      customView.spacingHorizontalLabel.text = it.toString()
      customView.horizontalLine.width = it
    }
    customView.horizontalLine.setColor(fillColor)

    customView.spacingVerticalSeek.max = 100
    customView.spacingVerticalSeek.onProgressChanged {
      customView.spacingVerticalLabel.text = it.toString()
      customView.verticalLine.width = it
    }
    customView.verticalLine.setColor(fillColor)

    customView.spacingHorizontalSeek.progress = horizontalSpacingPref.get()
    customView.spacingVerticalSeek.progress = verticalSpacingPref.get()

    return dialog
  }

  private fun notifyActivity() {
    val materialDialog = dialog as? MaterialDialog ?: return
    val customView = materialDialog.getCustomView() ?: return

    val horizontal = customView.spacingHorizontalSeek.progress
    horizontalSpacingPref.set(horizontal)
    val vertical = customView.spacingVerticalSeek.progress
    verticalSpacingPref.set(vertical)

    context?.onSpacingChanged(
        horizontal = horizontal,
        vertical = vertical
    )
  }

  override fun onAttach(context: Context?) {
    super.onAttach(context)
    this.context = context as? SpacingCallback
  }
}
